### ISSUE TITLE
Reloading DbContext for updated entity

### ISSUE THREAD

**Issue by APIWT** [artiomchi/FlexLabs.Upsert#59]
Hey @artiomchi!

I am having some trouble with stale data sticking around in the DbContext, and I was hoping you could maybe help me. Take the following example:

```
public class User
{
    public long Id { get; set; }

    public IEnumerable<Pet> Pets { get; set; }
}

public class Pet
{
    public long Id { get; set; }

    public long UserId { get; set; }

    public User User { get; set; }

    public uint Age { get; set; }
}
```

and imagine the following code:
```
// pretend that before this line the user that we are about to load has a pet with the age of 5
// load the user from the database
var user = await dbContext.User.Include(u => u.Pets).FirstOfDefaultAsync(u => u.Id == 123);

// the pet is now 7, just for an example
var somePet = user.Pet.FirstOrDefault();
somePet.Age = 7;

await this.context.Pet.Upsert(somePet)
    .On(v => new { p.Id })
    .WhenMatched(existingPet => new Pet
    {
        Age = somePet.Age,
    })
    .RunAsync();

// some time later, get the user back from the database / context
var updatedUser = await dbContext.User.Include(u => u.Pets).FirstOrDefaultAsync(u => u.Id == 123);

var samePet = updatedUser.Pet.FirstOrDefault();

Console.WriteLine(samePet.Age); // this line should print out 7 but it prints out 5 instead, which is the age of the pet from before upserting
```

I am able to work around the issue by calling `await dbContext.Entry(somePet).ReloadAsync()` right after the `Upsert` call.

Is this functionality intended? Is there a recommended pattern for avoiding this type of issue?

Thanks as always for your help!

**Comment by artiomchi** [comment]
Hey @APIWT 

That's... a hard one :)

Because of EF Core's complexity, I'm forced to operate outside of it's change tracker. In fact, the Upsert library creates SQL statements and just runs them on the SQL server using `dbContext.Database.ExecuteSqlCommand(string)`

Because of that, any changes that have happened on the server side aren't reflected in the change tracker in the current instance of the db context.

I have a plan though. Once I get a chance to work on #29 (and figure out some issues that are stopping me from doing it), I'll be able to load the values of changed entries back to code. With that in hand I'll be able to update the change tracker entities (which will also fix #44).

At the moment, using `ReloadAsync()` or creating a new context would be my recommended way of handling it, even though it's far from ideal or elegant

**Comment by APIWT** [comment]
Sounds good! I had a feeling this was just a side effect of manually executing SQL. Thank you very much for your help :) We will stick with reloading until you have some time to implement #29 etc
